kctf-operator/pkg/controller/challenge/challenge_controller.go

```diff
@@ -0,0 +1,268 @@
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a Challenge object and makes changes based on the state read
+// and what is in the Challenge.Spec
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileChallenge) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling Challenge")
+
+	// Fetch the Challenge instance
+	challenge := &kctfv1alpha1.Challenge{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, challenge)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			reqLogger.Info("Challenge resource not found. Ignoring since object must be deleted")
+			return reconcile.Result{}, nil
+		}
+
+		// Error reading the object - requeue the request.
+		reqLogger.Error(err, "Failed to get Challenge")
+		return reconcile.Result{}, err
+	}
+
+	// Check if the deployment already exists, if not create a new one
+	found := &appsv1.Deployment{}
+	err = r.client.Get(context.TODO(), types.NamespacedName{Name: challenge.Name,
+		Namespace: challenge.Namespace}, found)
+
+	// Set default values not configured by kubebuilder
+	SetDefaultValues(challenge)
+
+	// Just enters here if it's a new deployment
+	if err != nil && errors.IsNotFound(err) {
+		// Define a new deployment
+		dep := r.deploymentForChallenge(challenge)
+		reqLogger.Info("Creating a new Deployment", "Deployment.Namespace",
+			dep.Namespace, "Deployment.Name", dep.Name)
+		err = r.client.Create(context.TODO(), dep)
+
+		if err != nil {
+			reqLogger.Error(err, "Failed to create new Deployment", "Deployment.Namespace",
+				dep.Namespace, "Deployment.Name", dep.Name)
+			return reconcile.Result{}, err
+		}
+
+		// Deployment created successfully - return and requeue
+		return reconcile.Result{Requeue: true}, nil
+
+	} else if err != nil {
+		reqLogger.Error(err, "Failed to get Deployment")
+		return reconcile.Result{}, err
+	}
+
+	// Creates autoscaling object
+	// Checks if maxreplicas exists to see if it's enabled
+	// If enabled, it checks if it already exists
+	autoscalingFound := &autoscalingv1.HorizontalPodAutoscaler{}
+	err = r.client.Get(context.TODO(), types.NamespacedName{Name: challenge.Name,
+		Namespace: challenge.Namespace}, autoscalingFound)
+
+	if challenge.Spec.HorizontalAutoscaling.MaxReplicas != 0 && err != nil && errors.IsNotFound(err) {
```

Comment:
I was trying to understand what's wrong with these comparisons for a while now.

I think the solution for these would be to use pointers in challenge.Spec like this:

```
HorizontalPodAutoscalerSpec *autoscalingv1.HorizontalPodAutoscalerSpec
```

And then all you have to do is to check that it's not nil:
```
if challenge.Spec.HorizontalPodAutoscalerSpec != nil ...
```

The same goes for the comparisons inside SetDefaultValues():
```
if challenge.Spec.PodTemplate != nil {}
```

etc.